### PR TITLE
WINC-950: WMCO bundle upgrade testing

### PR DESCRIPTION
**Pull request by sebsoto**
Adds a job that upgrades from the WMCO released for the previous
OpenShift minor version, to the WMCO being tested. This is being marked
as not required. Once the job is fully functional, after changes to the
WMCO repo, this job will be made required and the already existing
upgrade job will be removed.

This job is split into 4 test steps
1) Install WMCO built for the previous OCP minor version.
2) Run initial tests against that WMCO using the
   `upgrade-test-setup` make target, creating Windows Nodes.
3) Upgrade WMCO to the version which is built from the code this job is
   testing, this causes those Windows nodes to be upgraded from the
   previous version to the current version.
4) Run tests using the `upgrade-test` make target, testing that the Node
   upgrade was successful.

### CODE REVIEW

**Comment by sebsoto** [comment]
/pj-rehearse pull-ci-openshift-windows-machine-config-operator-master-vsphere-e2e-upgrade

**Comment by sebsoto** [comment]
/pj-rehearse pull-ci-openshift-windows-machine-config-operator-master-vsphere-e2e-upgrade

**Comment by sebsoto** [comment]
/pj-rehearse pull-ci-openshift-windows-machine-config-operator-master-vsphere-e2e-upgrade

**Comment by sebsoto** [comment]
/pj-rehearse pull-ci-openshift-windows-machine-config-operator-master-vsphere-e2e-upgrade

**Comment by sebsoto** [comment]
/pj-rehearse pull-ci-openshift-windows-machine-config-operator-master-vsphere-e2e-upgrade

**Comment by sebsoto** [comment]
/pj-rehearse pull-ci-openshift-windows-machine-config-operator-master-vsphere-e2e-upgrade

**Comment by sebsoto** [comment]
/pj-rehearse pull-ci-openshift-windows-machine-config-operator-master-vsphere-e2e-upgrade

**Comment by sebsoto** [comment]
/pj-rehearse pull-ci-openshift-windows-machine-config-operator-master-vsphere-e2e-upgrade

**Comment by sebsoto** [comment]
/pj-rehearse pull-ci-openshift-windows-machine-config-operator-master-vsphere-e2e-upgrade

**Comment by sebsoto** [comment]
/pj-rehearse pull-ci-openshift-windows-machine-config-operator-master-vsphere-e2e-upgrade

**Comment by sebsoto** [comment]
/pj-rehearse pull-ci-openshift-windows-machine-config-operator-master-vsphere-e2e-upgrade

**Comment by sebsoto** [comment]
/pj-rehearse pull-ci-openshift-windows-machine-config-operator-master-vsphere-e2e-upgrade

**Comment by sebsoto** [comment]
/pj-rehearse pull-ci-openshift-windows-machine-config-operator-master-vsphere-e2e-upgrade

**Comment by sebsoto** [comment]
/pj-rehearse pull-ci-openshift-windows-machine-config-operator-master-vsphere-e2e-upgrade

**Comment by sebsoto** [comment]
/pj-rehearse pull-ci-openshift-windows-machine-config-operator-master-vsphere-e2e-upgrade

**Comment by sebsoto** [comment]
/pj-rehearse pull-ci-openshift-windows-machine-config-operator-master-aws-e2e-upgrade

**Comment by sebsoto** [comment]
/pj-rehearse pull-ci-openshift-windows-machine-config-operator-master-vsphere-e2e-upgrade

**Comment by sebsoto** [comment]
/pj-rehearse pull-ci-openshift-windows-machine-config-operator-master-vsphere-e2e-upgrade

**Comment by sebsoto** [comment]
/pj-rehearse pull-ci-openshift-windows-machine-config-operator-master-vsphere-e2e-operator

**Comment by sebsoto** [comment]
/pj-rehearse pull-ci-openshift-windows-machine-config-operator-master-vsphere-e2e-upgrade

**Comment by sebsoto** [comment]
/pj-rehearse pull-ci-openshift-windows-machine-config-operator-master-vsphere-e2e-upgrade

**Comment by sebsoto** [comment]
/pj-rehearse pull-ci-openshift-windows-machine-config-operator-master-vsphere-e2e-upgrade

**Comment by sebsoto** [comment]
/pj-rehearse pull-ci-openshift-windows-machine-config-operator-master-vsphere-e2e-upgrade

**Comment by sebsoto** [comment]
/pj-rehearse pull-ci-openshift-windows-machine-config-operator-master-vsphere-e2e-upgrade

**Comment by sebsoto** [comment]
/pj-rehearse pull-ci-openshift-windows-machine-config-operator-master-vsphere-e2e-upgrade

**Comment by sebsoto** [comment]
/pj-rehearse pull-ci-openshift-windows-machine-config-operator-master-vsphere-e2e-upgrade

**Comment by sebsoto** [comment]
/pj-rehearse pull-ci-openshift-windows-machine-config-operator-master-vsphere-e2e-upgrade

**Comment by sebsoto** [comment]
/pj-rehearse pull-ci-openshift-windows-machine-config-operator-master-vsphere-e2e-upgrade

**Comment by sebsoto** [comment]
/pj-rehearse pull-ci-openshift-windows-machine-config-operator-master-vsphere-e2e-upgrade

**Comment by sebsoto** [comment]
/pj-rehearse pull-ci-openshift-windows-machine-config-operator-master-vsphere-e2e-upgrade

File: ci-operator/config/openshift/windows-machine-config-operator/openshift-windows-machine-config-operator-master.yaml [hunk 1]
```diff
@@ -1,4 +1,12 @@
 base_images:
+  old-bundle:
```
Comment (aravindhp) [inline ci-operator/config/openshift/windows-machine-config-operator/openshift-windows-machine-config-operator-master.yaml:2]: Lets be specific here and name this `4.13-bundle` or `previous-bundle`. A follow up question is what happens when we move to 4.15? Do we have to manually fix this up?

File: ci-operator/config/openshift/windows-machine-config-operator/openshift-windows-machine-config-operator-master.yaml [hunk 2]
```diff
@@ -1,4 +1,12 @@
...
+    name: "4.13"
+    namespace: ocp
+    tag: windows-machine-config-operator-bundle
+  old-wmco:
```
Comment (aravindhp) [inline ci-operator/config/openshift/windows-machine-config-operator/openshift-windows-machine-config-operator-master.yaml:6]: Same comment here and elsewhere `old` is used.

File: ci-operator/config/openshift/windows-machine-config-operator/openshift-windows-machine-config-operator-master.yaml [hunk 3]
```diff
@@ -118,6 +126,75 @@ tests:
...
+    test:
+    - as: install
+      cli: latest
+      commands: |
+        export WMCO_NS=openshift-windows-machine-config-operator
+        oc create namespace $WMCO_NS
+        oc label --overwrite=true ns $WMCO_NS openshift.io/cluster-monitoring=true \
+        pod-security.kubernetes.io/enforce=privileged
+        operator-sdk run bundle --timeout=10m --security-context-config restricted -n $WMCO_NS "$OLD_BUNDLE" \
+        || oc get csv -n $WMCO_NS |awk {'print $1'} | tail -n1 | xargs oc patch csv -n $WMCO_NS --type='json' -p="[{\"op\": \"replace\", \"path\": \"/spec/install/spec/deployments/0/spec/template/spec/containers/0/image\", \"value\": \"$OLD_OPERATOR\"}]"
```
Comment (aravindhp) [inline ci-operator/config/openshift/windows-machine-config-operator/openshift-windows-machine-config-operator-master.yaml:142]: Please add a comment here stating what this command is doing.

**Review by aravindhp — CHANGES_REQUESTED** [review]
Thanks for working on this, @sebsoto. Please address my comment.

File: ci-operator/config/openshift/windows-machine-config-operator/openshift-windows-machine-config-operator-master.yaml [hunk 4]
```diff
@@ -118,6 +126,75 @@ tests:
...
+        sleep 10
+        oc delete deployment -n $WMCO_NS windows-machine-config-operator
```
[the remark this answers is not in the document]
Comment (sebsoto) [reply ci-operator/config/openshift/windows-machine-config-operator/openshift-windows-machine-config-operator-master.yaml:144]: Commented

**Comment by sebsoto** [comment]
/pj-rehearse pull-ci-openshift-windows-machine-config-operator-master-vsphere-e2e-upgrade

File: ci-operator/config/openshift/windows-machine-config-operator/openshift-windows-machine-config-operator-master.yaml [hunk 5]
```diff
@@ -118,6 +126,79 @@ tests:
...
+    - as: install
+      cli: latest
+      commands: |
+        export WMCO_NS=openshift-windows-machine-config-operator
+        oc create namespace $WMCO_NS
+        oc label --overwrite=true ns $WMCO_NS openshift.io/cluster-monitoring=true \
+        pod-security.kubernetes.io/enforce=privileged
+        # Deploy the bundle which will fail to come up due to it using the operator image via a tag that is not able to
+        # be made public. The csv needs to be patched to point the image via the tag that is pullable by the job.
+        operator-sdk run bundle --timeout=10m --security-context-config restricted -n $WMCO_NS "$PREVIOUS_BUNDLE" \
```
Comment (saifshaikh48) [inline ci-operator/config/openshift/windows-machine-config-operator/openshift-windows-machine-config-operator-master.yaml:143]: will this hit the 10 min timeout every time?

File: ci-operator/config/openshift/windows-machine-config-operator/openshift-windows-machine-config-operator-master.yaml (hunk 4 above)
[the remark this answers is not in the document]
Comment (saifshaikh48) [reply ci-operator/config/openshift/windows-machine-config-operator/openshift-windows-machine-config-operator-master.yaml:148]: I though the deployment automatically gets deleted + re-created by olm after patching the csv? Or is this done to speed up the process

File: ci-operator/config/openshift/windows-machine-config-operator/openshift-windows-machine-config-operator-master.yaml (hunk 5 above)
Comment (aravindhp) [reply ci-operator/config/openshift/windows-machine-config-operator/openshift-windows-machine-config-operator-master.yaml:143]: Can't you patch the CSV first?

File: ci-operator/config/openshift/windows-machine-config-operator/openshift-windows-machine-config-operator-master.yaml [hunk 6]
```diff
@@ -118,6 +126,79 @@ tests:
...
+        || oc get csv -n $WMCO_NS |awk {'print $1'} | tail -n1 | xargs oc patch csv -n $WMCO_NS --type='json' \
+        -p="[{\"op\": \"replace\", \"path\": \"/spec/install/spec/deployments/0/spec/template/spec/containers/0/image\", \"value\": \"$PREVIOUS_OPERATOR\"}]"
+        sleep 10
+        # Delete the deployment which will then be recreated by the subscription controller with the correct image
+        oc delete deployment -n $WMCO_NS windows-machine-config-operator
+        sleep 5m
```
Comment (aravindhp) [inline ci-operator/config/openshift/windows-machine-config-operator/openshift-windows-machine-config-operator-master.yaml:149]: Why do you need the sleep given the wait below?

File: ci-operator/config/openshift/windows-machine-config-operator/openshift-windows-machine-config-operator-master.yaml [hunk 7]
```diff
@@ -118,6 +126,79 @@ tests:
...
+      commands: |
+        export WMCO_NS=openshift-windows-machine-config-operator
+        oc create namespace $WMCO_NS
+        oc label --overwrite=true ns $WMCO_NS openshift.io/cluster-monitoring=true \
+        pod-security.kubernetes.io/enforce=privileged
+        # Deploy the bundle which will fail to come up due to it using the operator image via a tag that is not able to
+        # be made public. The csv needs to be patched to point the image via the tag that is pullable by the job.
+        operator-sdk run bundle --timeout=10m --security-context-config restricted -n $WMCO_NS "$PREVIOUS_BUNDLE" \
+        || oc get csv -n $WMCO_NS |awk {'print $1'} | tail -n1 | xargs oc patch csv -n $WMCO_NS --type='json' \
+        -p="[{\"op\": \"replace\", \"path\": \"/spec/install/spec/deployments/0/spec/template/spec/containers/0/image\", \"value\": \"$PREVIOUS_OPERATOR\"}]"
```
Comment (aravindhp) [inline ci-operator/config/openshift/windows-machine-config-operator/openshift-windows-machine-config-operator-master.yaml:145]: Where does `$PREVIOUS_BUNDLE` and `$PREVIOUS_OPERATOR` get set?

File: ci-operator/config/openshift/windows-machine-config-operator/openshift-windows-machine-config-operator-master.yaml (hunk 5 above)
Comment (sebsoto) [reply ci-operator/config/openshift/windows-machine-config-operator/openshift-windows-machine-config-operator-master.yaml:143]: The timeout will not be hit as there is an issue in operator-sdk which will not respect the timeout parameter correctly.  https://github.com/operator-framework/operator-sdk/pull/6440

File: ci-operator/config/openshift/windows-machine-config-operator/openshift-windows-machine-config-operator-master.yaml (hunk 4 above)
[the remark this answers is not in the document]
Comment (sebsoto) [reply ci-operator/config/openshift/windows-machine-config-operator/openshift-windows-machine-config-operator-master.yaml:148]: This is to speed up the process

File: ci-operator/config/openshift/windows-machine-config-operator/openshift-windows-machine-config-operator-master.yaml (hunk 6 above)
Comment (sebsoto) [reply ci-operator/config/openshift/windows-machine-config-operator/openshift-windows-machine-config-operator-master.yaml:149]: Good question. I dont remember and don't think its actually required. Removing and hoping it works in the pj-rehearse

File: ci-operator/config/openshift/windows-machine-config-operator/openshift-windows-machine-config-operator-master.yaml [hunk 8]
```diff
@@ -109,6 +117,79 @@ tests:
...
+        requests:
+          cpu: 1000m
+          memory: 400Mi
+    - as: upgrade
+      cli: latest
+      commands: |
+        export WMCO_NS=openshift-windows-machine-config-operator
+        operator-sdk run bundle-upgrade --timeout=10m --security-context-config restricted -n $WMCO_NS "$OO_BUNDLE"
+        sleep 5m
+        oc wait --timeout=5m --for condition=Available -n openshift-windows-machine-config-operator deployment windows-machine-config-operator
```
Comment (aravindhp) [inline ci-operator/config/openshift/windows-machine-config-operator/openshift-windows-machine-config-operator-master.yaml:171]:
```suggestion
        oc wait --timeout=5m --for condition=Available -n $WMCO_NS deployment windows-machine-config-operator
```

**Review by aravindhp — APPROVED** [review]

**Comment by sebsoto** [comment]
/pj-rehearse pull-ci-openshift-windows-machine-config-operator-master-vsphere-e2e-upgrade

**Comment by sebsoto** [comment]
/pj-rehearse pull-ci-openshift-windows-machine-config-operator-master-vsphere-e2e-upgrade

**Pull request merged**